### DISCUSSION TITLE
Async support

### DISCUSSION THREAD

**Discussion by tomasol** [GregoryConrad/rearch-rs#21] [Q&A] [answered]
How does this approach compose with async/await, e.g. in the axum example, how can the the read function `list_todos_capsule` race with a timeout?

**Comment by GregoryConrad** [comment, marked as answer]
Hi!👋 

(Sorry for the late reply; I’ve been on vacation and super busy)

Getting the todos here is a synchronous operation since redb currently only supports sync operations. Regardless, my initial approach would likely be to [`spawn_blocking`](https://docs.rs/tokio/latest/tokio/task/index.html#spawn_blocking) the list_todos call in a new task and then run a timer in another task that uses the JoinHandle of the first task and cancels it if the time is up. Then, you can use that Future from the combination of those two tasks in the async fn used for Axum.

probably isn’t relevant for this question, but you can also take a look at the rearch-tokio crate also present in this repo. Consider those side effects alpha quality for now, but I plan on polishing them soon. That can give an idea on how you can interop with async code in ReArch itself (instead of outside of it and with Axum directly).
